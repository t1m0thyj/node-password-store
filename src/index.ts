import * as fs from "fs";
import * as os from "os";
import * as path from "path";

export class PasswordStore {
    private kbpgp: any;
    private keyManager: any;
    private passwordStoreDir: string;

    private constructor(private privateKeyFile: string, passwordStoreDir?: string) {
        this.kbpgp = require("kbpgp");
        this.passwordStoreDir = passwordStoreDir ?? path.join(os.homedir(), ".password-store");
    }

    public static async create(privateKeyFile: string, passwordStoreDir?: string): Promise<PasswordStore> {
        const instance = new PasswordStore(privateKeyFile, passwordStoreDir);
        if (fs.existsSync(instance.privateKeyFile)) {
            return new Promise((resolve, reject) => {
                instance.kbpgp.KeyManager.import_from_armored_pgp({
                    armored: fs.readFileSync(instance.privateKeyFile, "utf-8")
                }, (err: Error | null, key: any) => {
                    if (err) return reject(err);
                    instance.keyManager = key;
                    resolve(instance);
                });
            });
        } else {
            // TODO Prompt for private key passphrase?
            return new Promise((resolve, reject) => {
                instance.kbpgp.KeyManager.generate_rsa({
                    userid: os.userInfo().username
                }, (err: Error | null, key: any) => {
                    if (err) return reject(err);
                    instance.keyManager = key;
                    key.sign({}, (err: Error | null) => {
                        if (err) return reject(err);
                        key.export_pgp_private({}, (err: Error | null, pgpPrivate: string) => {
                            if (err) return reject(err);
                            fs.writeFileSync(instance.privateKeyFile, pgpPrivate, { mode: 0o600 });
                            key.export_pgp_public({}, (err: Error | null, pgpPublic: string) => {
                                if (err) return reject(err);
                                fs.writeFileSync(instance.privateKeyFile + ".pub", pgpPublic, { mode: 0o644 });
                                resolve(instance);
                            });
                        });
                    });
                });
            });
        }
    }

    public getPassword(service: string, account: string): Promise<string | null> {
        const passwordFile = path.join(this.passwordStoreDir, service, account) + ".gpg";
        if (fs.existsSync(passwordFile)) {
            const ring = new this.kbpgp.keyring.KeyRing();
            ring.add_key_manager(this.keyManager);
            return new Promise((resolve, reject) => {
                this.kbpgp.unbox(
                    { keyfetch: ring, raw: fs.readFileSync(passwordFile) },
                    (err: Error | null, literals: any) => {
                        if (err) return reject(err);
                        resolve(literals[0].toString().slice(0, -1));
                    }
                );
            });
        } else {
            return Promise.resolve(null);
        }
    }

    public setPassword(service: string, account: string, password: string): Promise<void> {
        const passwordFile = path.join(this.passwordStoreDir, service, account) + ".gpg";
        fs.mkdirSync(path.dirname(passwordFile), { mode: 0o700, recursive: true });
        return new Promise((resolve, reject) => {
            this.kbpgp.box({
                msg: Buffer.from(password + "\n"),
                encrypt_for: this.keyManager
            }, (err: Error | null, _: string, result: Buffer) => {
                if (err) return reject(err);
                resolve(fs.writeFileSync(passwordFile, result, { mode: 0o600 }));
            });
        });
    }

    public deletePassword(service: string, account: string): Promise<boolean> {
        // TODO Verify key before deletion?
        const passwordFile = path.join(this.passwordStoreDir, service, account) + ".gpg";
        if (fs.existsSync(passwordFile)) {
            return new Promise((resolve, reject) => {
                fs.rm(passwordFile, (err: Error | null) => {
                    if (err) return reject(err);
                    resolve(true);
                })
            });
        } else {
            return Promise.resolve(false);
        }
    }

    public findCredentials(service: string): Promise<{ account: string; password: string }[]> {
        const serviceDir = path.join(this.passwordStoreDir, service);
        const credentials: { account: string; password: string }[] = [];
        return new Promise((resolve, reject) => {
            fs.readdir(serviceDir, (err: Error | null, files: string[]) => {
                if (err) return reject(err);
                Promise.all(files.filter((file) => file.endsWith(".gpg"))
                    .map((file) => {
                        const account = file.slice(0, -4);
                        return this.getPassword(service, account)
                            .then((password) => {
                                if (password) credentials.push({ account, password });
                            });
                    })
                ).then(() => resolve(credentials));
            });
        });
    }

    public findPassword(service: string): Promise<string | null> {
        return this.getPassword("", service);
    }
}
