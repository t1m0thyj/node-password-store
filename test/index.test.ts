import { PasswordStore } from "../src";

describe("password-store", () => {
    let passApi: PasswordStore;

    beforeAll(() => {
        passApi = PasswordStore.create("test.pgp", ".password-store");
    }, 60000);

    it("get/setPassword with ASCII string", async () => {
        await passApi.setPassword("TestKeytar", "TestASCII", "ASCII string");
        const password = await passApi.getPassword("TestKeytar", "TestASCII");
        expect(password).toBe("ASCII string");
    });

    it("get/setPassword with mixed character set", async () => {
        await passApi.setPassword("TestKeytar", "TestCharSet", "I 💔 ASCII");
        const password = await passApi.getPassword("TestKeytar", "TestCharSet");
        expect(password).toBe("I 💔 ASCII");
    });

    it("get/setPassword with UTF-16 chars", async () => {
        await passApi.setPassword("TestKeytar", "TestUTF16", "🌞🌙🌟🌴");
        const password = await passApi.getPassword("TestKeytar", "TestUTF16");
        expect(password).toBe("🌞🌙🌟🌴");
    });

    it("get/setPassword with CJK symbols", async () => {
        await passApi.setPassword("TestKeytar", "TestCJK", "「こんにちは世界」");
        const password = await passApi.getPassword("TestKeytar", "TestCJK");
        expect(password).toBe("「こんにちは世界」");
    });

    it("findCredentials prints credentials in test service and verifies member names", async () => {
        const credentials = await passApi.findCredentials("TestKeytar");
        expect(credentials.length).toBe(4);
        expect(credentials[0].account).toBeDefined();
        expect(credentials[0].password).toBeDefined();
        expect(credentials.sort((a, b) => a.account.localeCompare(b.account))).toMatchSnapshot();
    });

    it("findPassword for ASCII string", async () => {
        const password = await passApi.findPassword("TestKeytar/TestASCII");
        expect(password).toBe("ASCII string");
    });

    it("findPassword for mixed character set", async () => {
        const password = await passApi.findPassword("TestKeytar/TestCharSet");
        expect(password).toBe("I 💔 ASCII");
    });

    it("findPassword for UTF-16", async () => {
        const password = await passApi.findPassword("TestKeytar/TestUTF16");
        expect(password).toBe("🌞🌙🌟🌴");
    });

    it("findPassword for CJK symbols", async () => {
        const password = await passApi.findPassword("TestKeytar/TestCJK");
        expect(password).toBe("「こんにちは世界」");
    });

    it("deletePassword deletes all test credentials", async () => {
        let result = await passApi.deletePassword("TestKeytar", "TestASCII");
        expect(result).toBe(true);
        result = await passApi.deletePassword("TestKeytar", "TestCharSet");
        expect(result).toBe(true);
        result = await passApi.deletePassword("TestKeytar", "TestUTF16");
        expect(result).toBe(true);
        result = await passApi.deletePassword("TestKeytar", "TestCJK");
        expect(result).toBe(true);
    });
});
