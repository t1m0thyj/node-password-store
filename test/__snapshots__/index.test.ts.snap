// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`password-store findCredentials prints credentials in test service and verifies member names 1`] = `
[
  {
    "account": "TestASCII",
    "password": "ASCII string",
  },
  {
    "account": "TestCharSet",
    "password": "I 💔 ASCII",
  },
  {
    "account": "TestCJK",
    "password": "「こんにちは世界」",
  },
  {
    "account": "TestUTF16",
    "password": "🌞🌙🌟🌴",
  },
]
`;
